src/controllers/controller.bar.js

```diff
@@ -235,27 +245,76 @@ module.exports = function(Chart) {
 			var me = this;
 			var scale = me.getIndexScale();
 			var options = scale.options;
+			var ticksOffset = options.ticks.offset;
+			var barThickness = options.barThickness;
 			var stackCount = me.getStackCount();
-			var fullSize = scale.isHorizontal() ? scale.width : scale.height;
-			var tickSize = fullSize / scale.getTicks().length;
-			var categorySize = tickSize * options.categoryPercentage;
-			var fullBarSize = categorySize / stackCount;
-			var barSize = fullBarSize * options.barPercentage;
+			var datasetIndex = me.index;
+			var length = me.getMeta().data.length;
+			var data = [];
+			var ruler = [];
+			var isHorizontal = scale.isHorizontal();
+			var min = isHorizontal ? scale.left : scale.top;
+			var max = min + (isHorizontal ? scale.width : scale.height);
+			var minInterval = Infinity;
+			var prev = -Infinity;
+			var i, curr, leftSampleSize, rightSampleSize, leftCategorySize, rightCategorySize, fullBarSize, barSize;
+
+			// First pass: store data pixels and calculate the minimum interval
+			for (i = 0; i < length; ++i) {
+				curr = scale.getPixelForValue(null, i, datasetIndex);
+				data.push(curr);
+				minInterval = Math.min(minInterval, curr - prev);
+				if (ticksOffset) {
```

Comment:
We need to handle an edge case like that a dataset contains only one data point that is on the edge of a scale, the scale has only one tick and `ticks.offset: false`. In this case, we cannot calculate the bar width based on a tick interval or the distance from the nearest edge (because it is on the edge), so it uses the distance from the farthest edge.